pyfixest/estimation/feols_.py

```diff
@@ -943,6 +944,114 @@ def wald_test(self, R=None, q=None, distribution="F"):
 
         return res
 
+    def bootstrap(
+        self,
+        reps: int,
+        seed: Optional[int] = None,
+        inplace: bool = False,
+        fit_func: Optional[Callable] = None,
+        **kwargs,
```

Comment:
would it make sense to offer a 'compressed' flag for the bootstrap where the bootstrap replicates are (optionally) run in compressed fashion? would speed things up a lot.

Reply:
Yes,agreed this would be cool to have. From the compressed version, we would sample from a multinomial to reflect the weights? Maybe there even is a smart way to resample from the compressed data that is identical to the uncompressed bootstrap that I'm not seeing?

Reply:
I was thinking more along the lines of each bootstrap iteration involving (take uncompresseddata > sample (either IID or cluster) > compress > call compressed regression) as in duckreg. 
With polars, I'd assume the sample+compress step is fast enough to make this effective? If not, we can punt on this.

Reply:
I think / hope that that is what is already implemented? The function argument `fit_func` allows to pass a custom "evaluation function" to the bootstrap, and I think that I pass `feols_compressed` with the "long" data set. So we resample from the long data -> compress -> fit on the compressed data using sufficient statistics. But maybe I am still misunderstanding your suggestion? 🤔 

There's a few tricks that we could still apply, i.e. we could pre-compute $A_{g} = inv(X_{g}'X_{g})$ where $X_{g}$ contains all observations from cluster g, where X is already compressed. There is one caveat - I think that $A_{g}$ might turn out difficult to compute when there are cluster fixed effects. 

Then we'd only have to recompute $X_{g} y_{g\text{mean}}$ for each cluster that is drawn in the boostrap and can compute the "overall" coefficient as 

$$
\hat{\beta} = \sum_{g \in E} A_{g} \times X_{g} y_{\text{mean}}
$$

where $E$ are all indices of the resampled clusters, i.e. $E = {1, 2, 2, 3, 4, 8, 9}$. 

If we **do not** use the compression strategy, we could even pre-compute $\times X_{g} y_{g}$ for each g before running the bootstrap, with both $X$ and $y$ in "long" format. 

Key ideas are from [this MacKinnon paper](https://www.econstor.eu/bitstream/10419/247206/1/qed-wp-1465.pdf), equation (12) is the key equation for us: 
![Uploading image.png…]()